tensorflow_quantum/datasets/spin_system.py

```diff
@@ -320,3 +320,250 @@ def tfi_chain(qubits, boundary_condition="closed", data_dir=None):
+        - `fidelity`: Python `float` overlap between the circuit state
+            and the exact ground state from exact diagonalization.
+        - `params`: Dict with Python `str` keys and Numpy`float` values.
+            Contains $M \times P $ parameters. Here $M$ is the number of
+            parameters per circuit layer and $P$ the circuit depth.
+        - `var_circuit`: Variational `cirq.Circuit` quantum circuit with
+            unresolved Sympy parameters.
+    """
+
+    supported_n = [4, 8, 12, 16]
+    supported_bc = ["closed"]
+    if any(isinstance(q, list) for q in qubits):
+        raise TypeError("qubits must be a one-dimensional list")
+
+    if not all(isinstance(q, cirq.GridQubit) for q in qubits):
+        raise TypeError("qubits must be a list of cirq.GridQubit objects.")
+
+    nspins = len(qubits)
+    depth = nspins // 2
+    if nspins not in supported_n:
+        raise ValueError("Supported number of spins are {}, received {}".format(
+            supported_n, nspins))
+
+    if boundary_condition not in supported_bc:
+        raise ValueError(
+            "Supported boundary conditions are {}, received {}".format(
+                supported_bc, boundary_condition))
+
+    data_path = _download_spin_data('XXZ_chain', boundary_condition, nspins,
+                                    data_dir)
+
+    name_generator = unique_name()
+
+    # 4 * N/2 parameters.
+    symbol_names = [next(name_generator) for _ in range(2 * nspins)]
+    symbols = [sympy.Symbol(name) for name in symbol_names]
+
+    # Define the circuit.
+    circuit = cirq.Circuit(cirq.X.on_each(qubits))
+    circuit.append(cirq.H(qubits[i]) for i in range(0, nspins, 2))
+    circuit.append(
+        cirq.CNOT(qubits[j], qubits[j + 1]) for j in range(0, nspins - 1, 2))
+    for d in range(depth):
+        circuit.append(
+            cirq.ZZ(qubits[j], qubits[j + 1])**(symbols[d])
+            for j in range(1, nspins - 1, 2))
+        circuit.append(
+            cirq.YY(qubits[j], qubits[j + 1])**(symbols[d + depth])
+            for j in range(1, nspins - 1, 2))
+        circuit.append(
+            cirq.XX(qubits[j], qubits[j + 1])**(symbols[d + depth])
+            for j in range(1, nspins - 1, 2))
+        if boundary_condition == "closed":
+            circuit.append(cirq.ZZ(qubits[-1], qubits[0])**(symbols[d]))
+            circuit.append(cirq.YY(qubits[-1], qubits[0])**(symbols[d + depth]))
+            circuit.append(cirq.XX(qubits[-1], qubits[0])**(symbols[d + depth]))
+        circuit.append(
+            cirq.ZZ(qubits[j], qubits[j + 1])**(symbols[d + 2 * depth])
+            for j in range(0, nspins - 1, 2))
+        circuit.append(
+            cirq.YY(qubits[j], qubits[j + 1])**(symbols[d + 3 * depth])
+            for j in range(0, nspins - 1, 2))
+        circuit.append(
+            cirq.XX(qubits[j], qubits[j + 1])**(symbols[d + 3 * depth])
+            for j in range(0, nspins - 1, 2))
```

Comment:
Repeating the `for j in range(1, nspins-1, 2)` loop over and over makes me wonder if we can't do something like:

```python
interacting_qubits = zip(qubits, qubits[1:])
if boundary_condition == 'closed':
  interacting_qubits.append((qubits[0], qubits[-1]))

for d in range(depth):
  for a, b in ineracting_qubits:
    circuit.append(cirq.XX(a, b))
    ....
    ....
    ....


```

Reply:
I tried this approach, but I am doing something wrong:
```python
even_qubits = zip([qubits[j] for j in range(0, nspins - 1, 2)],
                  [qubits[j + 1] for j in range(0, nspins - 1, 2)])
for q1,q2 in even_qubits:
   circuit.append(cirq.ZZ(q1,q2)**(symbols[d + 2 * depth]))
   circuit.append(cirq.YY(q1,q2)**(symbols[d + 3 * depth]))
   circuit.append cirq.XX(q1,q2)**(symbols[d + 3 * depth]))
```
gives the wrong fidelities and energies, whereas
```python
for j in range(0, nspins - 1, 2):
    circuit.append(cirq.ZZ(qubits[j], qubits[j + 1])**(symbols[d + 2 * depth]))
    circuit.append(cirq.YY(qubits[j], qubits[j + 1])**(symbols[d + 3 * depth]))
    circuit.append(cirq.XX(qubits[j], qubits[j + 1])**(symbols[d + 3 * depth]))
```
works correctly. What am I missing here?

Reply:
I'm not at a computer to verify this, but I would guess that if the code breaks only when you substitute these two sections in and out that it has to do with the zip. I think it might be the fact that the zip state is producing a generator expression. Doing this:

```python3
even_qubits = list(zip(...))
```

Might fix things ?

Reply:
That did the trick.